crates/matrix-sdk-crypto/src/identities/room_identity_state.rs

```diff
@@ -0,0 +1,1078 @@
+            if let Ok(user_id) = user_id {
+                // Ignore non-existent users
+                if let Some(user_identity) = self.room.user_identity(user_id).await {
+                    match event.content.membership {
+                        MembershipState::Join | MembershipState::Invite => {
+                            if let Some(update) = self.update_user_state(user_id, &user_identity) {
+                                return vec![update];
+                            }
+                        }
+                        MembershipState::Leave | MembershipState::Ban => {
+                            let leaving_state = state_of(&user_identity);
+                            if leaving_state == IdentityState::PinViolation {
+                                // If a user with bad state leaves the room, set them to Pinned,
+                                // which effectively removes them
+                                return vec![self.set_state(user_id, IdentityState::Pinned)];
+                            }
+                        }
+                        MembershipState::Knock => {
+                            // No need to do anything when someone is knocking
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+
+        // We didn't find a relevant update, so return an empty list
+        vec![]
+    }
+
+    fn update_user_state(
+        &mut self,
+        user_id: &UserId,
+        user_identity: &UserIdentity,
+    ) -> Option<IdentityStatusChange> {
+        if let UserIdentity::Other(_) = &user_identity {
+            // If the user's state has changed
+            let new_state = state_of(user_identity);
+            let old_state = self.known_states.get(user_id);
+            if new_state != old_state {
+                Some(self.set_state(user_identity.user_id(), new_state))
+            } else {
+                // Nothing changed
+                None
+            }
+        } else {
+            // Ignore updates to our own identity
+            None
+        }
+    }
+
+    fn set_state(&mut self, user_id: &UserId, new_state: IdentityState) -> IdentityStatusChange {
+        // Remember the new state of the user
+        self.known_states.set(user_id, &new_state);
+
+        // And return the update
+        IdentityStatusChange { user_id: user_id.to_owned(), changed_to: new_state }
+    }
+}
+
+fn state_of(user_identity: &UserIdentity) -> IdentityState {
+    if user_identity.is_verified() {
+        IdentityState::Verified
+    } else if user_identity.has_verification_violation() {
+        IdentityState::PreviouslyVerified
+    } else if let UserIdentity::Other(u) = user_identity {
+        if u.identity_needs_user_approval() {
+            IdentityState::PinViolation
+        } else {
+            IdentityState::Pinned
+        }
+    } else {
+        IdentityState::Pinned
```

Comment:
I wonder how we should consider our own identity...
We designed that based on the requirement that you must verify your own devices or you are locked out of the app if not, so basically our own identity is never a problem.

But if we let you have insecure device, I am not sure what to do here.

Reply:
Since this is driven by the need to display banners in the UI, and nothing will be automatically displayed when the current user's device is insecure, changes to our own identity are suppressed.

Reply:
Discussed IRL and we will address this over the next few weeks.